pkg/index/job/readreplica/rotate/service/rotator.go

```diff
@@ -0,0 +1,335 @@
+// Copyright (C) 2019-2023 vdaas.org vald team <vald@vdaas.org>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package service
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	"github.com/vdaas/vald/internal/errors"
+	client "github.com/vdaas/vald/internal/k8s/client"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/observability/trace"
+	"github.com/vdaas/vald/internal/sync/errgroup"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	apiName = "vald/index/job/readreplica/rotate"
+)
+
+// Rotator represents an interface for indexing.
+type Rotator interface {
+	Start(ctx context.Context) error
+}
+
+type rotator struct {
+	namespace           string
+	volumeName          string
+	readReplicaLabelKey string
+	readReplicaID       string
+	client              client.Client
+	listOpts            client.ListOptions
+}
+
+// New returns Indexer object if no error occurs.
+func New(replicaID string, opts ...Option) (Rotator, error) {
+	r := new(rotator)
+
+	if replicaID == "" {
+		return nil, fmt.Errorf("readreplica id is empty. it should be set via MY_TARGET_REPLICA_ID env var")
+	}
+	r.readReplicaID = replicaID
```

Comment:
It would be better to use functional options e.g. `func() WithReplicaID(id string) error {}`

Reply:
Because `replicaID` is mandatory. Let's merge this first and discuss it later.